src/libstd/sys_common/os_str_bytes.rs

```diff
@@ -106,9 +106,20 @@ impl Buf {
 
     #[inline]
     pub fn as_slice(&self) -> &Slice {
+        // Safety: Slice just wraps [u8],
+        // and &*self.inner is &[u8], therefore
+        // transmuting &[u8] to &Slice is safe.
         unsafe { mem::transmute(&*self.inner) }
     }
 
+    #[inline]
+    pub fn as_mut_slice(&mut self) -> &mut Slice {
```

Comment:
One additional safety concern for the internal API: isn't `OsStr` not allowed to contain NULs? assuming that's the case, getting a `&mut [u8]` to it's internals should be `unsafe`.

Reply:
OsStr is a superset of UTF-8, so I don't see how this can be true.

Reply:
Ah, ok. nevermind.

Reply:
Are there any invariants on maintaining valid WTF-8 on Windows though?

Reply:
@BurntSushi On Windows, `Slice` wraps `Wtf8` and, to the best of my understanding, no methods exist on `Slice` that could invalidate the underlying Wtf8. `Slice` does declare the inner `Wtf8` public, but the `Wtf8` does NOT declare the inner `[u8]` public. So I don't see any way that Wtf8 invariant could be broken. I'll add a couple of comments about upholding any platform-specific encoding, however.

The goal of this is to enable `make_ascii_lowrcase` and `make_ascii_uppercase` for `OsStr`. It is less clear to me that such methods wouldn't break Wtf8, so certainly if there is a problem there please point it out!

Reply:
Comments sound great, thanks! (I'm not too familiar with this code, so I was mostly shooting from the hip. :-))